src/main/scala/beam/agentsim/infrastructure/ScaleUpCharging.scala

```diff
@@ -129,90 +135,123 @@ trait ScaleUpCharging extends {
+            val varianceRange: Double =
+              listRemainingRange.map(range => range - meanRangeInMeters).map(t => t * t).sum / numObservation
+            val vehicleTypes =
+              ObservedVehicleTypes(
+                vehicleType,
+                vehicleAlias,
+                reservedFor,
+                numObservation,
+                meanRangeInMeters,
+                varianceRange
+              )
+            new CPair[ObservedVehicleTypes, java.lang.Double](vehicleTypes, vehicleTypes.numObservation.toDouble)
           }
           .toVector
-        val vehicleInfoSummary = VehicleInfoSummary(
-          numObservation = numObservation,
-          totPowerInKW = totPowerInKW,
-          totDurationInSec = totDurationInSec,
-          meanDuration = meanDur,
-          varianceDuration = varianceDur,
-          meanSOC = meanSOC,
-          varianceSOC = varianceSOC,
-          meanEnergy = meanEnergy,
-          varianceEnergy = varianceEnergy,
-          new EnumeratedDistribution[VehicleTypeInfo](mersenne, pmfVehicleTypeInfo.asJava),
-          new EnumeratedDistribution[String](mersenne, pmfActivityType.asJava)
-        )
-        parkingActivityType -> (data, vehicleInfoSummary)
-      })
-      .flatMap { case (tazId, activityType2vehicleInfo) =>
-        val parkingInquiriesTriggers = Vector.newBuilder[ScheduleTrigger]
-        activityType2vehicleInfo.foldLeft((0.0, 0.0, Vector.empty[CPair[ParkingActivityType, java.lang.Double]])) {
-          case ((powerAcc, numEventsAcc, pmfAcc), (parkingActivityType, (_, dataSummary))) =>
-            val scaleUpFactor = scaleUpFactors.getOrElse(parkingActivityType, defaultScaleUpFactor) - 1
-            val power = scaleUpFactor * dataSummary.totPowerInKW
-            val pmf = new CPair[ParkingActivityType, java.lang.Double](parkingActivityType, power)
-            val numEvents = scaleUpFactor * dataSummary.numObservation
-            (powerAcc + power, numEventsAcc + numEvents, pmfAcc :+ pmf)
-        } match {
-          case (totPowerInKWToSimulate, totNumberOfEvents, pmf) if totPowerInKWToSimulate > 0 =>
-            val distribution = new EnumeratedDistribution[ParkingActivityType](mersenne, pmf.asJava)
+        (pmfObservedActivities, pmfObservedVehicleTypes)
+      }
+      .flatMap {
+        case (tazId, (pmfObservedActivities, pmfObservedVehicleTypes)) if tazTreeMap.getTAZ(tazId).isDefined =>
+          val parkingInquiriesTriggers = Vector.newBuilder[ScheduleTrigger]
+          val activitiesLocationInCurrentTAZ: Map[String, Vector[ActivityLocation]] =
+            activitiesLocationMap
+              .get(tazId)
+              .map(_.filterNot { a =>
+                allPersonsWhichCarIsChargingByTAZ.getOrElse(tazId, mutable.HashSet.empty).contains(a.personId) ||
+                allVirtualPersonsByTAZ.synchronized(allVirtualPersonsByTAZ(tazId).contains(a.personId))
```

Comment:
Using `synchronized` may impact performance. I'd suggest using `java.util.concurrent.ConcurrentHashMap` instead of mutable.HashSet. In this case `synchronized` is not needed.

Reply:
Done